### DISCUSSION TITLE
Autonomous Emergency braking query

### DISCUSSION THREAD

**Discussion by hardikbvaghela** [prl-mushr/mushr#82] [Q&A]
hello. my robot is working with slam algorithm. but my plan is to start AEB in mushr so how can I do this? is there any reference or anything is available, that can help me to create AEB in my mushr. 
Thank you.

**Comment by schmittlema** [comment]
What is AEB? Autonomous Emergency Brake?

**Reply by hardikbvaghela** [reply]
Yes Autonomous emergency braking system.

**Reply by schmittlema** [reply]
The[ software system overview](https://mushr.io/tutorials/overview/) discusses the mux a bit. Different control messages are sent to the mux from various processes (autonomous, teleop, safety) and the mux decides based on priority which should actually be executed. Safety (`../input/safety`) should have a higher priority than autonomous. So your AEB system should publish stop commands to that topic whenever it would like to stop the vehicle.

How you go about deciding when to stop the vehicle is up to you. You could use the slam map but that relies on the slam map being accurate. You could also use the lidar directly and stop the vehicle whenever the robot gets too close to any lidar points. That would be the simple solution and likely safest

**Reply by hardikbvaghela** [reply]
hello
actually when i am doing autonomous navigation that time if i run mushr_pf and mushr_rhc node but only mushr_pf is initializing but mushr_rhc not i dont know why both are not running togrther. then i am installing all mushr_rhc dependency but there a error with some file is missing. 

[
![Screenshot from 2022-07-24 15-24-43](https://user-images.githubusercontent.com/93780099/180649134-0a54fa31-1618-427b-80cf-579ae78a5c88.png)
](url)

**Reply by schmittlema** [reply]
You may be able to get some insight from this discussion: https://github.com/prl-mushr/mushr/discussions/75

We have been updating the stack/nav and are in the testing stage to make sure everything is working as expected.

**Reply by hardikbvaghela** [reply]
hello
now everything is working. but I have another idea. if I run my Mushr autonomously in the room. and if the front of the small dummy like a pedestrian is moving right to left so is it possible to measure pedestrian movement or pedestrian velocity or speed with the help of 2d lidar point cloud data?

I can attach one picture so you can understand what I want to ask you.

Thank you.
![mushr](https://user-images.githubusercontent.com/93780099/190139870-14a2d058-2e38-40bd-a7e0-18f35d09e224.PNG)

**Reply by schmittlema** [reply]
Yes this should be possible but likely easier if you use lidar and camera. First you need to [detect the person](https://developer.nvidia.com/blog/realtime-object-detection-in-10-lines-of-python-on-jetson-nano/), this is done best by using a person detection neural network using the camera. Once you know the person's position in the image, you can look at the stereo or lidar at that point to detect their depth. This is also likely easier with depth images as you don't need to convert from the camera frame to the lidar frame. Once you have the person's 3d position  you simply need to see how that position changes over time and convert that to a velocity vector (just subtract current position - previous position and divide by the tme difference). One thing to keep in mind is that if the robot is moving, the position of the person will be different without them moving so you need to correct for the change in robot position from each detection. This can be done with the odometry `/car/odom` or a global position provided by the particle filter. So get detection, correct for robot movement, compute velocity from previously corrected person positions.
